proposals/NNNN-specialize.md

```diff
@@ -0,0 +1,193 @@
+
+The `@specialize` attribute can be placed on any generic function. It takes an argument with the same syntax as a `where` clause for a generic signature. 
+
+The `where` clause must fully specialize all the generic placeholders of the types in the function signature. In the case of a protocol extension, as seen above, that includes specifying `Self`.
+
+It can also be used on extensions of generic types, on computed properties (note, it must be put on `get` explicitly, not on the shorthand where it is elided), and on free functions:
+
+```swift
+extension Array where Element: BinaryInteger {
+  @specialize(where Element == Int)
+  func sum() -> Double {
+    reduce(0) { $0 + Double($1) }
+  }
+  
+  var product: Double {
+    @specialize(where Element == Int8)
+    get { reduce(1) { $0 * Double($1) } }
+  }
+}
+
+@specialize(where T == Int)
+func sum<T: BinaryInteger>(_ numbers: T...) -> Double {
+  numbers.reduce(0) { $0 + Double($1) }
+}
+```
+
+All placeholders must be fully specified in the `where` clause even if they do not appear to be used in the function body:
+
+```swift
+extension Dictionary where Value: BinaryInteger {
+  // error: Too few generic parameters are specified in '_specialize' attribute (got 1, but expected 2)
+  // note: Missing constraint for 'Key' in 'specialize' attribute
+  @specialize(where Value == Int)
+  func sum() -> Double {
+    values.reduce(0) { $0 + Double($1) }
+  }
+}
+```
+
+Bear in mind that even when not explicitly used in the body, they may be used implicitly. For example, in calculating the location of stored properties. Depending on how `Dictionary` is laid out, it may be important to know the size of the `Key` even if key values aren't used. But even if there is absolutely no use of a generic type in the body being specialized, the type must be explicitly specified. This requirement could be loosened in future, see "Partial Specialization" in future directions.
+
+Where multiple placeholders need to be specified separately, they can be separated by commas, such as in the fix for the above example:
+
+```swift
+extension Dictionary where Value: BinaryInteger {
+  @specialize(where Value == Int, Key == Int)
+  func sum() -> Double {
+    values.reduce(0) { $0 + Double($1) }
+  }
+}
+```
+
+## Source compatibility
+
+The addition or removal of explicit specializations has no impact on the caller, and is opt-in. As such it has no source compatability implications.
+
+## ABI compatibility
+
+This proposal covers only _internal_ specializations, dispatched to from within the implementation of an unspecialized generic function. As such it has no impact on ABI. It can be applied to existing ABI-stable functions in libraries built for distribution, and can be removed later without ABI impact. Generic functions can continue to be inlinable to be specialized by the caller, and the code to dispatch to specialized versions will not appear in the inlinable code emitted into the swift interface file.
```

Comment:
Is there a use case for a function both being inlinable and having a specialize attribute?

Reply:
Yes. It means the specialization is created for invocations via e.g. an existential, while still allowing _other_ types to be specialized from concrete call sites (especially types outside the module).